crates/test-programs/build.rs

```diff
@@ -0,0 +1,192 @@
+    let mut out =
+        File::create(out_dir.join("wasi_tests.rs")).expect("error generating test source file");
+    build_tests("wasi-tests", &out_dir).expect("building tests");
+    test_directory(&mut out, "wasi-tests", &out_dir).expect("generating tests");
+}
+
+fn build_tests(testsuite: &str, out_dir: &Path) -> io::Result<()> {
+    let mut cmd = Command::new("cargo");
+    cmd.args(&[
+        "build",
+        "--release",
+        "--target=wasm32-wasi",
+        "--target-dir",
+        out_dir.to_str().unwrap(),
+    ])
+    .stdout(Stdio::inherit())
+    .stderr(Stdio::inherit())
+    .current_dir(testsuite);
+    let output = cmd.output()?;
+
+    let status = output.status;
+    if !status.success() {
+        panic!(
+            "Building tests failed: exit code: {}",
+            status.code().unwrap()
+        );
+    }
+
+    Ok(())
+}
+
+fn test_directory(out: &mut File, testsuite: &str, out_dir: &Path) -> io::Result<()> {
+    let mut dir_entries: Vec<_> = read_dir(out_dir.join("wasm32-wasi/release"))
+        .expect("reading testsuite directory")
+        .map(|r| r.expect("reading testsuite directory entry"))
+        .filter(|dir_entry| {
+            let p = dir_entry.path();
+            if let Some(ext) = p.extension() {
+                // Only look at wast files.
+                if ext == "wasm" {
+                    // Ignore files starting with `.`, which could be editor temporary files
+                    if let Some(stem) = p.file_stem() {
+                        if let Some(stemstr) = stem.to_str() {
+                            if !stemstr.starts_with('.') {
+                                return true;
+                            }
+                        }
+                    }
+                }
+            }
+            false
+        })
+        .collect();
+
+    dir_entries.sort_by_key(|dir| dir.path());
+
+    writeln!(
+        out,
+        "mod {} {{",
+        Path::new(testsuite)
+            .file_stem()
+            .expect("testsuite filename should have a stem")
+            .to_str()
+            .expect("testsuite filename should be representable as a string")
+            .replace("-", "_")
+    )?;
+    writeln!(out, "    use super::{{runtime, utils, setup_log}};")?;
+    for dir_entry in dir_entries {
+        write_testsuite_tests(out, dir_entry, testsuite)?;
+    }
+    writeln!(out, "}}")?;
+    Ok(())
+}
+
+fn write_testsuite_tests(out: &mut File, dir_entry: DirEntry, testsuite: &str) -> io::Result<()> {
+    let path = dir_entry.path();
+    let stemstr = path
+        .file_stem()
+        .expect("file_stem")
+        .to_str()
+        .expect("to_str");
+
+    writeln!(out, "    #[test]")?;
+    if ignore(testsuite, stemstr) {
+        writeln!(out, "    #[ignore]")?;
+    }
+    writeln!(
+        out,
+        "    fn r#{}() -> anyhow::Result<()> {{",
+        &stemstr.replace("-", "_")
+    )?;
+    writeln!(out, "        setup_log();")?;
+    write!(
```

Comment:
Since we're moving the file, we can probably fix this to be `writeln!` instead of `write!` (it's something I noticed when I was running tests under a debugger and it opened the generated source).